documents/system-design-document.md

```diff
@@ -0,0 +1,455 @@
+
+  For detailed project requirements, please refer to the [Product Requirements Document (PRD)](https://github.com/what-a-cool-team/docs/blob/main/prd/what-a-todo.md).
+
+## 2. Functional Requirements
+| Priority | Requirements |
+|----------|--------------|
+| P0 | Users are able to create an account |
+| P0 | Users are able to delete their account |
+| P0 | Users are able to log in securely |
+| P0 | Users are able create new tasks with a name, and optionally add description and tags |
+| P0 | Users are able to update existing tasks to change details such as name, description, status, and tags|
+| P1 | Users are able to view a list of all their tasks |
+| P2 | Users are able to delete tasks they no longer need |
+| P2 | Users are able to search tasks by name or tags |
+| P2 | Users are able to filter tasks by status |
+
+## 3. Operational requirements
+  * **Security**: 
+    * The system should grant users access only to the tasks that they own.
+    * The system should allow users to log in and out using their registered ID and password.
+    * The system should allow users to delete their accounts.
+    * Administrators should be restricted from accessing user passwords.
+    
+## 4. Domain Model Design
+
+### 4.1 Entities
+|Entity| ID example |
+|------|--------------|
+|User|
+|Task|
+|Tag|
+|Status| "Created", "Committed", "In Progress", "Completed" |
+
+
+### 4.2 Attributes
+
+| Entity | Question | Logical type | Example value | 
+| -------- | ------- | ---------------| ------------- | 
+|User| What is the account ID of this user? | string  | "user"| 
+|Task| What is the name of this task? | string | "Grocery shopping" | 
+|Task| What is the description for this task?" | string | "1. Meat,<br> 2. Onion,"| 
+| Task | What is the status of this task? | enum | "In Progress" |
+| Tag | What is the name of this tag? | string | "Daily" |
+
+### 4.3 Links
+|Entity * Entity| Sentences | Cardinality |
+|---------------|-----------|-------------|
+|User < Task|User creates many Tasks<br> Task is created by *only one* User|1:N|
+|Task = Tag |Task can have *several* tags assigned<br> Tag can be assigned to *several* tasks |M:N |
+
+### Domain Model Diagram
+![alt text](diagram.png)
+
+## 5. Tables
+ 
+```sql
+CREATE TABLE users (
+  id SERIAL PRIMARY KEY,
+  account_id VARCHAR(64) NOT NULL
+);
+
+CREATE TABLE tasks (
+  id SERIAL PRIMARY KEY,
+  user_id INT NOT NULL,
+  name VARCHAR(128) NOT NULL,
+  description VARCHAR(128),
+  status VARCHAR(20) NOT NULL CHECK (status IN ('Created', 'Committed', 'In Progress', 'Completed')),
+  FOREIGN KEY (user_id) REFERENCES users(id)
+);
+
+
+CREATE TABLE tags (
+  id SERIAL PRIMARY KEY,
+  name VARCHAR(64) NOT NULL
+);
+
+CREATE TABLE tasks_tags (
+  task_id INTEGER NOT NULL,
+  tag_id INTEGER NOT NULL,
+  PRIMARY KEY (task_id, tag_id),
+  KEY (tag_id)
+  FOREIGN KEY (task_id) REFERENCES tasks(id),
+  FOREIGN KEY (tag_id) REFERENCES tags(id)
+);
+```
+
+## 6. API Document
+```yaml
+POST /api/users/:
```

Comment:
nit: Can we make the API endpoint pattern consistent? This has the ending slash, but below the `POST /api/tasks` does not, and again `GET /api/tasks/` does. Let's make it consistent, and I'd suggest without ending slash.